BCDA-2282 Feature: Implemented Patient all endpoint

tbellj opened the pull request:
Fixes [BCDA-2282](https://jira.cms.gov/browse/BCDA-2282)

Problem:
In order to be more FHIR compliant, we needed to change our bulk data requests to match the query pattern for bulk Patient requests. This new pattern matches FHIR standards, see example here: /api/v1/Patient/$export?_type where _$export?_type_ will be followed by a comma delimited list of FHIR resource types.

Previously, requests were in the form: "/api/v1/ExplanationOfBenefit/$export"
With this current change it'll look like: "/api/v1/Patient/$export?_type=ExplanationOfBenefit"

Proposed Changes:
Modified all bulk data requests to match the FHIR standard for bulk Patient requests with the inclusion of query parameters to list different FHIR resource types. Also modified 429 logic to fit with this new workflow. This change encompasses both BCDA-2282 and BCDA-2283 work.

Change Details:
bcda/models/SwaggerStructs.go - added resource type parameter to query
bcda/database/connection.go - returns a queue database connection
bcda/models/models.go - added a resource type string field to the job key table
bcda/responseutils/writer.go - modified bulk requests urls to match new query pattern
bcda/web/api.go - modified bulk patient requests to match new query pattern, also changed 429 logic to work with these new changes
bcda/web/api_test.go - modified to work with the new bulk patient requests pattern
bcda/web/router.go - removed eob and coverage api routes
bcda/web/router_test.go - modified to work with the new bulk patient requests pattern, also removed unnecessary feature flags for outdated coverage and patient endpoint exposure.
bcdaworker/main.go - added a resource type to be saved with the job key
docker-compose.yml - added "sslmode=disable" to the queue database url
test/performance_test/performance.go - modified to work with the new bulk patient requests pattern
test/postman_test/BCDA-SSAS.postman_collection.json - modified to work with the new bulk patient requests pattern
test/postman_test/BCDA_Tests.postman_collection.json - modified to work with the new bulk patient requests pattern
test/postman_test/BCDA_Tests_Sequential.postman_collection.json - modified to work with the new bulk patient requests pattern
test/smoke_test/bcda_client.go - modified to work with the new bulk patient requests pattern

Security Implications:
Yes, this change does deal with PII/PHI. We are changing the way our users will receive sensitive data via the new query patterns for bulk patient requests outlined in the bulk FHIR spec. See here: http://hl7.org/fhir/uv/bulkdata/STU1/OperationDefinition-patient-export.html.

<!-- Does the change deal with PII/PHI at all? What should reviewers look for in
terms of security concerns? -->

- [ ] new software dependencies

<!-- If yes, list the new dependencies and briefly note any relevant security impacts -->

- [ ] security controls or supporting software altered

<!-- If yes, what security controls or supporting software are affected? -->

- [ ] new data stored or transmitted

<!-- If yes, what new data are we storing or transmitting? Is the data considered PII/PHI? -->

- [x] security checklist is completed for this change
https://confluence.cms.gov/pages/viewpage.action?pageId=228920304

<!-- If yes, provide a link to the security checklist in Confluence here. -->

- [ ] requires more information or team discussion to evaluate security implications

<!-- Use this to indicate you're unsure how this change may impact system security
and would like to solicit the team's feedback. Optionally, provide background
information regarding your questions and concerns. -->

Acceptance Validation
Yes, all tests pass.

![image](https://user-images.githubusercontent.com/42976557/69373177-3d72e880-0c71-11ea-964c-2e9a79a5aae1.png)
---------------------------------------------------------------------------------------------------
![image](https://user-images.githubusercontent.com/42976557/69444309-17575200-0d1e-11ea-9820-1e3687a31004.png)

---------------------------------------------------------------------------------------------------
Swagger screenshots:
![image](https://user-images.githubusercontent.com/42976557/70072031-38446080-15c4-11ea-856d-f7f4950d6598.png)

![image](https://user-images.githubusercontent.com/42976557/70072058-46927c80-15c4-11ea-8c04-f6eee9ea5e4d.png)


Feedback Requested:
Any and all

msnook commented on bcda/web/api.go:
```diff
@@ -33,31 +33,6 @@ import (
-		500: errorResponse
-*/
-
-func bulkEOBRequest(w http.ResponseWriter, r *http.Request) {
```
I think we are going to want to keep `/ExplanationOfBenefit` and `/Coverage` around for a little while, for backwards compatibility.

msnook commented on bcda/responseutils/writer.go:
```diff
@@ -81,7 +81,7 @@ func CreateCapabilityStatement(reldate time.Time, relversion, baseurl string) *f
 						Name: "export",
 						Definition: &fhirmodels.Reference{
-							Reference: baseurl + "/api/v1/ExplanationOfBenefit/$export",
+							Reference: baseurl + "/api/v1/Patient/$export?_type=ExplanationOfBenefit",
```
Why explicitly indicating the `_type`?  The Capability is the `/Patient` endpoint.  What is the FHIR-compliant way of defining this endpoint in a capability statement.

tbellj replied:
The spec doesn't really provide a "fhir compliant" way of showing this type of endpoint...at least i couldn't find it lol. But i also don't see why we wouldn't want to do this. (It's informing our users how to make the different calls we support via the new query parameters) But it's also possible I'm missing your point entirely. What is it specifically that concerns you??

msnook replied:
I guess the problem I see is that we should only need to include the endpoint we support (i.e., `/Patient`).  If we start adding all the `_type` options, then it opens up the door to tying us to include all permutations in the CapabilityStatement.

msnook commented on bcda/web/api.go:
```diff
@@ -204,6 +216,91 @@ func bulkRequest(t string, w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusAccepted)
 }
 
+func check429(jobs []models.Job, types []string, w http.ResponseWriter) ([]string, bool) {
```
This seems overly complex.  What if we give a `429` regardless of the type of job they have requested?
They have the freedom to request `/Patient` (ALL) and get everything with one request.

tbellj replied:
workin on that confluence page..stay tuned :)

msnook commented on test/smoke_test/bcda_client.go:
```diff
@@ -81,7 +81,7 @@ func getAccessToken() string {
 	client := &http.Client{}
 
-	var url string = fmt.Sprintf("%s://%s/api/v1/%s/$export", proto, apiHost, resourceType)
+	var url string = fmt.Sprintf("%s://%s/api/v1/Patient/$export?_type=%s", proto, apiHost, resourceType)
```
What about a test for `/Patient` all?

msnook replied:
Yeah, we'll need a test for `/Patient` all so we can validate that the response format is correct.

tbellj replied:
it was my understanding that we would add the new test cases (including patient all) in this future ticket here: BCDA-2284 and this pr was to retrofit all existing cases to work with the new patient query parameters. I could be wrong though, let me know if this is totally not the case.

msnook replied:
Yes, all permutations can be addressed in BCDA-2284.  However, the standard `/Patient` request should probably be tested as part of this ticket.

tbellj replied:
done!

msnook commented on bcda/database/connection.go:
```diff
@@ -38,3 +38,18 @@ func GetGORMDbConnection() *gorm.DB {
 	return db
 }
+
+func GetQueueDbConnection() *sql.DB {
```
See comment below on `429`.  Maybe we can avoid requiring a connection to the queue db.

tbellj replied:
good concerns, yeah I think putting that page together to explain the different scenarios will give us more of a chance for some good dialogue around this. I just NEED TO FINISH IT.

msnook commented on bcda/web/api.go:
```diff
@@ -279,9 +376,6 @@ func jobStatus(w http.ResponseWriter, r *http.Request) {
 			scheme = "https"
 		}
 
-		re := regexp.MustCompile(`/(ExplanationOfBenefit|Patient|Coverage)/\$export`)
```
Should we still have this check here, but for `/Patient` only?

tbellj replied:
Nah we don't really need it with these changes. Its really only there to create a regex for getting the resource type from the job's request url to properly display the fhir resources in the output of the payload. That now comes from the resource type being stored in the jobKey table that gets populated during construction of the individual files themselves. So now the resource type and the filename both get stored together and retrieved together with this output.

msnook commented:
It is not clear to me where the response format has changed to accommodate a `/Patient` all request?  
It should align with this format . (see `Complete` section): http://hl7.org/fhir/uv/bulkdata/STU1/export/index.html#endpoint-1

tbellj commented:
> It is not clear to me where the response format has changed to accommodate a `/Patient` all request?
> It should align with this format . (see `Complete` section): http://hl7.org/fhir/uv/bulkdata/STU1/export/index.html#endpoint-1

yup, fair enough. I should have added a screenshot for proof. Just added it.

msnook commented:
> > It is not clear to me where the response format has changed to accommodate a `/Patient` all request?
> > It should align with this format . (see `Complete` section): http://hl7.org/fhir/uv/bulkdata/STU1/export/index.html#endpoint-1
> 
> yup, fair enough. I should have added a screenshot for proof. Just added it.

Thanks, that response looks awesome.  Does it still work in scenario where there are multiple Patient, Coverage, and EOB files?  (This is probably the most common scenario).  Maybe we can test with ACO Medium or ACO Small?  Or, you could update the env vars locally so that each resource type only has like 20 beneficiaries per file (forcing multiple files locally)

tbellj commented:
> > > It is not clear to me where the response format has changed to accommodate a `/Patient` all request?
> > > It should align with this format . (see `Complete` section): http://hl7.org/fhir/uv/bulkdata/STU1/export/index.html#endpoint-1
> > 
> > 
> > yup, fair enough. I should have added a screenshot for proof. Just added it.
> 
> Thanks, that response looks awesome. Does it still work in scenario where there are multiple Patient, Coverage, and EOB files? (This is probably the most common scenario). Maybe we can test with ACO Medium or ACO Small? Or, you could update the env vars locally so that each resource type only has like 20 beneficiaries per file (forcing multiple files locally)

Added a screenshot showing the multiple (Patient, Coverage, EOB) files after changing the env vars to 20 benes.

msnook commented on test/smoke_test/bulk_data_requests.sh:
```diff
@@ -8,4 +8,6 @@ go run bcda_client.go -host=api:3000 -clientID=$CLIENT_ID -clientSecret=$CLIENT_
\ No newline at end of file
+go run bcda_client.go -host=api:3000 -clientID=$CLIENT_ID -clientSecret=$CLIENT_SECRET -endpoint=Coverage
+echo "Running Patient All"
+go run bcda_client.go -host=api:3000 -clientID=$CLIENT_ID -clientSecret=$CLIENT_SECRET -endpoint=All
```
`All` is technically not a FHIR-compliant resource type.  I know this is just a test script, but it may cause confusion. 
 What if we make `-endpoint` optional?  If it is not passed in, then we assume it means we want all resource types.

msnook replied:
Can we also add this to the Jenkinsfile for smoke tests, so that the default `/Patient` gets tested in AWS?

msnook commented on test/smoke_test/bcda_client.go:
```diff
@@ -80,8 +80,13 @@ func getAccessToken() string {
+	var url string
 
-	var url string = fmt.Sprintf("%s://%s/api/v1/%s/$export", proto, apiHost, resourceType)
+	if resourceType != "All" {
```
See comment below.

msnook commented on bcda/web/api.go:
```diff
@@ -33,37 +32,12 @@ import (
 	swagger:route GET /api/v1/Patient/$export bulkData bulkPatientRequest
 
-	Start patient data export
+	Start patient, explanation of benefit, and coverage data export
```
Maybe something like:
`Start data export for all supported resources types`

msnook replied:
From Swagger, I am able to make a call to `/Patient`.  I think we should add the ability for the user to supply a `_type` parameter so that users can execute export jobs for individual resource types.  Does Swagger support that?

msnook commented on bcda/web/api.go:
```diff
@@ -33,37 +32,12 @@ import (
+	Start patient, explanation of benefit, and coverage data export
 
-	Initiates a job to collect patient data from the Blue Button API for your ACO.
+	Initiates a job to collect patient, explanation of benefit, and coverage data from the Blue Button API for your ACO.
```
Maybe:
`Initiates a job to collect data from the Blue Button API for your ACO.  Support resource types are Patient, Coverage, and ExplanationOfBenefit.`

msnook commented on bcda/models/SwaggerStructs.go:
```diff
@@ -132,6 +132,15 @@ type FileParam struct {
+	// in: query
+	// style: form
+	// explode: false
+	ResourceType []string `json:"_type"`
```
Shouldn't this be`string` not `[]string`?
